File: rxlibrary/src/main/java/pers/cxd/rxlibrary/RxUtil.java
package pers.cxd.rxlibrary;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.ObservableTransformer;
import io.reactivex.rxjava3.core.Observer;
import io.reactivex.rxjava3.schedulers.Schedulers;

/**
 * Miscellaneous {@code RxJava} utility methods.
 *
 * @author pslilysm
 * @since 1.0.0
 */
public class RxUtil {

    public static <D> void execute(@NonNull Observable<D> observable, Observer<D> observer, ObservableTransformer<D, D> transformer) {
        observable.compose(transformer)
                .subscribe(observer);
    }

    public static class Transformers {

        private static final ObservableTransformer IOToMain = upstream ->
                upstream.subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread());

        private static final ObservableTransformer IO = upstream -> upstream.subscribeOn(Schedulers.io());

        private static final ObservableTransformer NON = upstream -> upstream;

        public static <T> ObservableTransformer<T, T> IOToMain() {
            return IOToMain;
        }

        public static <T> ObservableTransformer<T, T> IO() {
            return IO;
        }

        public static <T> ObservableTransformer<T, T> NON() {
            return NON;
        }

    }

}

File: rxlibrary/src/main/java/pers/cxd/rxlibrary/RxUtil.kt
package pers.cxd.rxlibrary

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.ObservableTransformer
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.schedulers.Schedulers

/**
 * Miscellaneous `RxJava` utility methods.
 *
 * @author pslilysm
 * @since 1.0.0
 */
object RxUtil {
    fun <D : Any> execute(
        observable: Observable<D>,
        observer: Observer<D>,
        transformer: ObservableTransformer<D, D>
    ) {
        observable.compose(transformer)
            .subscribe(observer)
    }

    object Transformers {
        private val IOToMain: ObservableTransformer<Any, Any> =
            ObservableTransformer<Any, Any> { upstream: Observable<Any> ->
                upstream.subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
            }
        private val IO: ObservableTransformer<Any, Any> =
            ObservableTransformer<Any, Any> { upstream: Observable<Any> -> upstream.subscribeOn(Schedulers.io()) }
        private val NON: ObservableTransformer<Any, Any> =
            ObservableTransformer<Any, Any> { upstream: Observable<Any> -> upstream }

        fun <T> IOToMain(): ObservableTransformer<T, T> {
            return IOToMain as ObservableTransformer<T, T>
        }

        fun <T> IO(): ObservableTransformer<T, T> {
            return IO as ObservableTransformer<T, T>
        }

        fun <T> NON(): ObservableTransformer<T, T> {
            return NON as ObservableTransformer<T, T>
        }
    }
}